internal/util/update/update_test.go

```diff
@@ -678,6 +680,137 @@ func TestCommand_Run_failInvalidPath(t *testing.T) {
 	}
 }
 
+func TestCommand_Run_badUpstreamLockRepo(t *testing.T) {
+	repos, w, clean := testutil.SetupReposAndWorkspace(t, map[string][]testutil.Content{
+		testutil.Upstream: {
+			{
+				Pkg: pkgbuilder.NewRootPkg().
+					WithResource(pkgbuilder.DeploymentResource),
+				Branch: mainBranch,
+			},
+		},
+	})
+	defer clean()
+
+	w.PackageDir = testPackageName
+	dir := "/"
+	kf := kptfileutil.DefaultKptfile(testPackageName)
+	kf.Upstream = &kptfilev1alpha2.Upstream{
+		Type: kptfilev1alpha2.GitOrigin,
+		Git: &kptfilev1alpha2.Git{
+			Repo:      repos[testutil.Upstream].RepoDirectory,
+			Directory: dir,
+			Ref:       mainBranch,
+		},
+		UpdateStrategy: kptfilev1alpha2.ResourceMerge,
+	}
+	kf.UpstreamLock = &kptfilev1alpha2.UpstreamLock{
+		Type: kptfilev1alpha2.GitOrigin,
+		GitLock: &kptfilev1alpha2.GitLock{
+			Repo:      "fake",
+			Directory: dir,
+			Ref:       mainBranch,
+		},
+	}
+	testutil.AddKptfileToWorkspace(t, w, kf)
+
+	// Update the local package
+	err := Command{
+		Pkg: pkgtest.CreatePkgOrFail(t, w.FullPackagePath()),
+	}.Run()
+
+	if assert.Error(t, err) {
+		assert.Contains(t, err.Error(), "does not appear to be a git repository")
+	}
+}
+
+func TestCommand_Run_badUpstreamLockRef(t *testing.T) {
+	repos, w, clean := testutil.SetupReposAndWorkspace(t, map[string][]testutil.Content{
+		testutil.Upstream: {
+			{
+				Pkg: pkgbuilder.NewRootPkg().
+					WithResource(pkgbuilder.DeploymentResource),
+				Branch: mainBranch,
+			},
+		},
+	})
+	defer clean()
+
+	w.PackageDir = testPackageName
+	dir := "/"
+	kf := kptfileutil.DefaultKptfile(testPackageName)
+	kf.Upstream = &kptfilev1alpha2.Upstream{
+		Type: kptfilev1alpha2.GitOrigin,
+		Git: &kptfilev1alpha2.Git{
+			Repo:      repos[testutil.Upstream].RepoDirectory,
+			Directory: dir,
+			Ref:       mainBranch,
+		},
+		UpdateStrategy: kptfilev1alpha2.ResourceMerge,
+	}
+	kf.UpstreamLock = &kptfilev1alpha2.UpstreamLock{
+		Type: kptfilev1alpha2.GitOrigin,
+		GitLock: &kptfilev1alpha2.GitLock{
+			Repo:      repos[testutil.Upstream].RepoDirectory,
+			Directory: dir,
+			Ref:       "fake",
+		},
+	}
+	testutil.AddKptfileToWorkspace(t, w, kf)
+
+	// Update the local package
+	err := Command{
+		Pkg: pkgtest.CreatePkgOrFail(t, w.FullPackagePath()),
+	}.Run()
+	if !assert.NoError(t, err) {
```

Comment:
This situation doesn't return an error? It is a bit of a strange state, but I would expect kpt to fail since the Ref doesn't exist.

Reply:
Looks like it tries to clone using `Commit` instead of `Ref` https://github.com/GoogleContainerTools/kpt/blob/786926a36f4a49872cb0fe986cc4a489ea67ab76/internal/util/update/update.go#L230

Reply:
Yeah, this is what it should do. We are using the `Commit` field to fetch the common ancestor for local and upstream so we can do a 3-way merge.
We don't actually use the `Ref` field in `GitLock`. I have been considering if it should be removed, but I haven't made a decision on that yet. We can just skip this test. We should test what happens if the `Commit` field has a value that doesn't exist in the repo.